src/libraries/icubmod/bcbBattery/bcbBattery.cpp

```diff
@@ -65,214 +80,111 @@ bool BcbBattery::open(yarp::os::Searchable& config)
-        do
-        {
-            output[0] = pSerial->receiveChar(serial_buff[0]);
-            sync0 = output[0] && (serial_buff[0] == '\0');
-            ++i;
-        } while (!(sync0) && (i < 20) && !isClosing);
-
-        if (isClosing)
-        {
-            return;
-        }
-
-        if (sync0)
-        {
-            output[1] = pSerial->receiveChar(serial_buff[1]); //voltage
-            output[2] = pSerial->receiveChar(serial_buff[2]); //voltage
-            output[3] = pSerial->receiveChar(serial_buff[3]); //current
-            output[4] = pSerial->receiveChar(serial_buff[4]); //current
-            output[5] = pSerial->receiveChar(serial_buff[5]); //charge
-            output[6] = pSerial->receiveChar(serial_buff[6]); //charge
-            output[7] = pSerial->receiveChar(serial_buff[7]); //status
-        }
-
-        i = 0;
-        do
-        {
-            output[8] = pSerial->receiveChar(serial_buff[8]);
-            syncr = output[8] && (serial_buff[8] == '\r');
-            if (syncr)
-            {
-                output[9] = pSerial->receiveChar(serial_buff[9]);
-                syncn = output[9] && (serial_buff[9] == '\n');
-
-                if (!syncn)
-                {
-                    if (!syncError && !silenceSyncWarnings)
-                    {
-                        yDebug() << "Sync error (n)";
-                    }
-                    syncError = true;
-                }
-            }
-            else
-            {
-                if (!syncError && !silenceSyncWarnings)
-                {
-                    yDebug() << "Sync error (r)";
-                }
-                syncError = true;
-            }
-            ++i;
-        } while (!(syncn) && (i < 20) && !isClosing);
-
-        if (isClosing)
-        {
-            return;
-        }
-
-        serial_buff[10] = 0;
+        recb = pSerial->receiveBytes(tmp_buff, buff_len);
     }
     else
     {
         yError("BcbBattery pSerial == NULL");
     }
-
-#if DEBUG_TEST
-    battery_voltage     = 40.0;
-    battery_current = 5.0;
-    battery_charge = 72.0;
-    battery_temperature = 35.0;
 #else
+    recb = test_buffer((unsigned char*)(tmp_buff), buff_len);
+#endif
 
     if (verboseEnable)
     {
-        char hexBuffer[31];
-        char decBuffer[41];
-        for (size_t i = 0; i < 10; ++i)
-        {
-            if (output[i])
-            {
-                sprintf(hexBuffer + 3*i,"%02X ", (unsigned int)(serial_buff[i] & 0xFF));
-                sprintf(decBuffer + 4*i,"%03u ", (unsigned int)(serial_buff[i] & 0xFF));
-            }
-            else
-            {
-                sprintf(hexBuffer + 3*i,"-- ");
-                sprintf(decBuffer + 4*i,"--- ");
-            }
-        }
-
-        yDebug("BcbBattery::run() serial_buffer is: (hex) %s, (dec) %s", hexBuffer, decBuffer);
+        printf("internal buffer:");
+        for (size_t i = 0; i < recb; i++)
+            printf("%02X ", (unsigned int)(tmp_buff[i] & 0xFF));
```

Comment:
Please, keep also the decimal printout, as it was before. This is useful to debug if the incoming values make sense without having to convert them from hexadecimal to decimal.

Reply:
well, the regex cannot fail, so a correct representation of the packet is always given in the subsequent hex representation of the packet  (line ~169 of the next commit) . In that debug print, decimal representation is no more useful since they match 1:1 with the values you can read from the interface...

Reply:
Still, we could have problems at the sender level. That has been useful on few occasions in the past.